### PR TITLE
rsx: FIFO processing rewrite

### PR DESCRIPTION
**Pull request by kd-11**
This PR essentially does 2 things:
1. Refactors FIFO processing into something more readable, including separation of the main loop from the flattener.
2. Significantly reworks how draw calls are processed. This part is still very unfinished, but the idea is to completely ignore the FIFO begin/end commands and use our own mechanisms to decide where to mark the scope of draw calls. The reorganization and optimization here results in quite a good fps boost in applications that are limited by RSX performance.

Other minor improvements include:
1. Fixes some minor bug in the surface_store that could return an empty set when asked to get an overlapping list with at least one confirmed hit.
2. Make the FIFO flattener engage dynamically depending on load to avoid slowdown when the extra processing overhead is not required.

At some point late in development, I scrapped the older system with the prebuffered queues and optimization passes. The performance hit of FIFO preprocessing was quite high and the prebuffering caused race conditions in poorly synchronized applications. This should be resolved in the final version, which has greatly simplified FIFO processing.

So, what does this all mean? Well, you could potentially get a ~10% boost in applications with an RSX bottleneck. This may not mean much for some applications, but the gains are significant at high draw call counts. Most other applications may only show a few fps gain.

### CODE REVIEW

File: rpcs3/Emu/RSX/RSXThread.cpp [hunk 1]
```diff
@@ -509,521 +636,55 @@ namespace rsx
...
+			m_decompiler_thread.reset();
 		}
 	}
 
-	void thread::on_exit()
+	std::string thread::get_name() const
 	{
-		m_rsx_thread_exiting = true;
+		return "rsx::thread";
+>>>>>>> rsx: Fixups
```
Comment (MSuih) [inline rpcs3/Emu/RSX/RSXThread.cpp]: Remove this.

Comment (kd-11) [reply rpcs3/Emu/RSX/RSXThread.cpp]: Yes, didn't catch that (I just grep for <<<<< and forgot the opposite. Looks like some definitions were moved on master.

Comment (kd-11) [reply rpcs3/Emu/RSX/RSXThread.cpp]: Ahh, the threading system was refactored recently. These variables no longer exist.

**Comment by MSuih** [comment]
Ratchet and Clank Trilogy: Performance is much, much worse. In Florana starting screen i get 0.9 FPS while master delivers 14.5 FPS. Strangely, the numbers shown on debug overlay are better despite the worse performance. [RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/2558113/RPCS3.log.gz)

Red Dead Redemption: Nothing is rendered. FPS changes but screen stays black. [RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/2558123/RPCS3.log.gz)

**Comment by kd-11** [comment]
I may have broken something during rebase. Investigating.

**Comment by legend800** [comment]
Captain America™: Super Soldier [BLUS30642] freezes soon after entering gameplay (this was the case on other WIP builds that I tried).  No perf gained (still ~17 fps in game). 

[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/2558234/RPCS3.log.gz)

WipEout® HD [NPUA80105] : Same thing. 

[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/2558242/RPCS3.log.gz)

**Comment by isJuhn** [comment]
@MSuih weird, I saw a 20% fps increase on the florana starting position

**Comment by kd-11** [comment]
Confirmed RDR regression, not sure if caused by botched rebase as this game was tested extensively before.

**Comment by darkdragonz12** [comment]
sin titulo
God of war 3

**Comment by legend800** [comment]
@kd-11 That may be true but just wanted to note that they all have regressed with this PR in the ways mentioned in the post.

**Comment by MSuih** [comment]
I don't have a save file which'd allow me to access beyond the first two maps, and getting anything done on this PR is a nightmare as the low fps causes extreme slowdown. But here are two screenshots which have very noticeable fps drops

image
(14.5 -> 0.9 FPS)
image
(30 -> 3,6 fps)

I have a RX 480, maybe this is AMD or Polaris specific regression?

**Comment by kd-11** [comment]
Likely, some of my work conflicts with earlier elad fix for u16 vs u32 index buffers. I'll investigate tomorrow, I have an AMD card on hand.

**Comment by kd-11** [comment]
Yea, that doesn't look right. On Nvidia, 7k draw calls in R&C give ~13fps. Maybe something regressed on AMD afterall.

**Comment by kd-11** [comment]
@MSuih R9 270x + R7 1700, Crimson 18.8.1
image

**Comment by kd-11** [comment]
I literally have ~170fps in R&C2 for that screenshot.

**Comment by kd-11** [comment]
I'll try and isolate it. The AMD team has been very responsive lately and we have some polaris workarounds that may need to be removed anyway.

**Comment by Xcedf** [comment]
Confirm GoW3 regression, but Tools of Destruction runs great with this PR and SPU LLVM

**Comment by Kravickas** [comment]
I would bump ''rsx'' version. Some games are black screen only with old cache.

**Comment by vietnam13231** [comment]
Infamous: no graphics after logos. Sound is going normally, but the game is unresponsive and FPS freezes. Some kind of purple error came out after exit, adding log.
[Infamous-RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/2559555/Infamous-RPCS3.log.gz)

NieR Replikant: only black and white screens are visible. Sound is going normally, FPS are refreshing, game is responsive.

ETERNAL SONATA: mostly black screen, sometimes with some object (like crystal in main menu) or effect (green wave in logos). Sound is going normally, FPS are refreshing, game is responsive.
 
P5, Hyperdimension Neptunia, Hatsune Miku: Project Diva F & F 2nd, Drakengard 3 works the same as in master.

**Comment by DanteBC** [comment]
Shadow of the colossus HD shows only black screen, although music is playing an the game is responsive
[SotCHD.RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/2559739/SotCHD.RPCS3.log.gz)
Neverdead also shows nothing except hud and subtitles ingame
Master:
neverdead
This PR
neverdead-pr

**Comment by Marthgun** [comment]
Tales of Xilia has a slight fps boost in Vulkan, but there are the occasional weird pop-ins, like drawing is hanging up, some places where it would lag slightly are all but gone, dips only a few fps except for one of the beginning areas in town.  overall feels more stable, not sure how much help this is, but here's the log

rpcs3-v0.0.5-2e3212d9_win64

[RPCS3.zip](https://github.com/RPCS3/rpcs3/files/2560373/RPCS3.zip)

**Comment by kd-11** [comment]
Rebase issue has been resolved, retest for graphical glitches.

**Comment by Anty0** [comment]
Building this PR on Arch Linux results in:
```
In file included from /mnt/AntyInternalSSD/RPCS3/rpcs3/rpcs3/Emu/RSX/RSXThread.h:11,
                 from /mnt/AntyInternalSSD/RPCS3/rpcs3/rpcs3/Emu/RSX/GSRender.h:3,
                 from /mnt/AntyInternalSSD/RPCS3/rpcs3/rpcs3/Emu/System.cpp:18:
/mnt/AntyInternalSSD/RPCS3/rpcs3/rpcs3/Emu/RSX/RSXFIFO.h:6:10: fatal error: Utilities/thread.h: No such file or directory
 #include <Utilities/thread.h>
```

Can be fixed by replacing with <Utilities/Thread.h>. (Actual filename is Thread.h not thread.h.)

**Comment by kd-11** [comment]
There may be more linux issues; haven't tested anything on linux yet. Will definitely fix whatever issues are left, because I think there may be more than just that include.

**Comment by Hasster1** [comment]
InFamous after loading freezes (showing black screen, and playing music.). If you will try to click the exit button, rpcs3 will crash.

**Comment by tabnk** [comment]
Tekken 6 and Tekken Tag Tournament 2.

Both games are less stalling and slightly more stable framerate.

**Comment by Emulator-Team-2** [comment]
[92fc2cdd ] 
AMD-RX 570

[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/2562117/RPCS3.log.gz)


RDR
```
F {rsx::thread} RSX: class std::runtime_error thrown: Verification failed: 
(in file c:\projects\rpcs3\rpcs3\emu\rsx\rsxfifo.cpp:156)
F {rsx::thread} RSX: class std::runtime_error thrown: Unreachable
(in file c:\projects\rpcs3\rpcs3\emu\rsx\rsxfifo.cpp:437)
F {rsx::thread} RSX: class std::runtime_error thrown: Verification failed: 
(in file c:\projects\rpcs3\rpcs3\emu\rsx\rsx_methods.cpp:1209)
```
[1RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/2562121/1RPCS3.log.gz)

**Comment by elfbk201** [comment]
Heavy rain - emulator crash.

{rsx::thread} RSX: Unimplemented block size in texels for texture format: 0x10
{rsx::thread} RSX: Unimplemented block size in bytes for texture format: 0x10

[RPCS3.log.zip](https://github.com/RPCS3/rpcs3/files/2562170/RPCS3.log.zip)

**Comment by MSuih** [comment]
@Emulator-Team-2 What's the FPS like? Exploding vertices are not an issue, they happen on AMD cards when RSX is pushed too far.

**Comment by Emulator-Team-2** [comment]
@MSuih On this screenshot, I had more than 13 FPS.

**Comment by darkdragonz12** [comment]
gow 3 and RDR returns to normal again
sin titulo

**Comment by vsub** [comment]
Persona 5 sees some speed boost but I can't say by how much...the fps keeps changing while I am not moving between 21-29 but on this pr it changes between 26-30

On a better cpu(6700HQ),the speed boost will probably be higher

**Comment by RainbowCookie32** [comment]
Red Dead Redemption: Game of the Year Edition [BLES01294] crashes on the main menu
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/2562903/RPCS3.log.gz)

**Comment by legend800** [comment]
NFL Blitz gfx fixed on latest commit. Stability issues remain though.

**Comment by vietnam13231** [comment]
P5: no changes on my PC.
Infamous: same as Hasster1
ETERNAL SONATA: Problem fixed, max FPS (not counting small dives).
NieR Replikant: Back to normal...ish? Plays the same as in master, but crashes whole emulator after exit (though that may be because of SPU LLVM).

**Comment by HexyFify** [comment]
improves fps in lbp1 by around 5-7fps

**Comment by MsDarkLow** [comment]
Ni No Kuni: Wrath of the White Birch shows flashing colors
[RPCS3NNKFifo.log.gz](https://github.com/RPCS3/rpcs3/files/2564250/RPCS3NNKFifo.log.gz)
[RPCS3NNKMaster.log.gz](https://github.com/RPCS3/rpcs3/files/2564251/RPCS3NNKMaster.log.gz)

PR
1pr
Master
2master

[EPILEPSY WARNING]
This is how horrible it looks: https://youtu.be/gG2cpJ-TVXY

**Comment by cesarnox** [comment]
Persona 5
intel i5 8400 integrated HD 630 

OPENGL
Master: 15-16.6fps
PR: 20fps

Vulkan
Master: working 
PR: Crash after some intro scenes, keep sound of the train for a while(almost every game that I test crash like this)

Version vulkan 1.1.85

**Comment by kd-11** [comment]
@MSuih Can you try running the default rpcs3 samples and report on what the performance is like? I have no issues with AMD on both linux and windows. Also, I did not encounter that cmake build error on linux, I wonder if its fixed now since travis also seems to pass just fine.

**Comment by Anty0** [comment]
@kd-11 I also didn't encounter @MSuih 's cmake problem on my machine (Arch Linux). Only the one I reported (which is now fixed).

**Comment by MSuih** [comment]
Yeah it's building now, I just had to manually reload the cmake project as it wasn't picking up the changes without that. As for the samples, here are the results:

Test | This PR | Master
-----|---------|--------
gs_gcm_basic_triangle | ~3000 fps | ~2700 fps
gs_gcm_cube | ~2830 fps | ~2850 fps
gs_gcm_handle_system_cmd | ~3160 fps | ~3160 fps
gs_gcm_hellow_world | ~3120 fps | ~3140 fps
gs_gcm_tetris | ~2500 fps | ~2500 fps

(Tested on R5 1600 @ 3,7 GHz and RX 480 8GB, driver version 18.10.2)

No changes, really. R&C Triology must be doing something unusual which upsets the drivers. Here's a video of the performance I get in case someone wants to laught at it: https://streamable.com/mrvcx

Also, did something break with Qt fullscreen in this PR? I'm getting horrible aliasing in all of the games I quickly tested with the latest version of this PR, but not on master. Setting resolution scaling to 50% still produces blurry image, so at least that works.

**Comment by kd-11** [comment]
Didn't touch anything Qt related, but I'll certainly take a look if I can identify other applications affected by the Ni No Kuni issue reported here.

**Comment by alexandrbril** [comment]
Resident Evil Code Veronica X HD
FPS boost
Master
bandicam 2018-11-10 21-09-36-149 mp4_snapshot_00 41_1
PR
bandicam 2018-11-10 21-14-36-041 mp4_snapshot_00 37_1
FPS 7-17->17-30
PC Spec: Intel Xeon E5 2696V3(12C/24T@3.77GHz), Geforce GTX 1050Ti, RAM 16Gb@2239

**Comment by vietnam13231** [comment]
FPS drop in P5 for me, from 20~30 to 14~25.

**Comment by darkdragonz12** [comment]
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/2568767/RPCS3.log.gz)

Metal Gear Rising FIFO ERROR INFINITE

E {rsx::thread} RSX: FIFO error: possible desync

**Comment by Parthorisian** [comment]
[Conan 2007 [BLUS30069] glitches](https://streamable.com/t9cw8)

**Comment by Kravickas** [comment]
Flickering glitches with this PR in Uncharted too, Master is ok.
https://streamable.com/x2hv9

**Comment by MSuih** [comment]
Should have probably tested this sooner, but for some reason I didn't think about it at all: The Polaris R&C performance issues only affect Vulkan, Opengl hasn't regressed and is actually running slightly better with this PR (10.4 -> 13.7 fps).

**Comment by kd-11** [comment]
@MsDarkLow Retest

**Comment by MsDarkLow** [comment]
> @MsDarkLow Retest

Can confirm the issue has been fixed. Very nice!

**Comment by kd-11** [comment]
@Parthorisian That has been 'fixed' in the newer builds. Same issue as Ni No Kuni.

**Comment by legend800** [comment]
Stability is better on new build (Captain America seems fine). And while Wipeout HD can get through a race now, it encounters a FIFO error at end of race ('possible desync' error message spammed).

2018-11-11 09_16_02-window

[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/2569600/RPCS3.log.gz)

**Comment by Xcedf** [comment]
Infamous demo now runs faster about 8-10% than master, but now before the moment i may control my character, i get Verification fail error in rsx_methods.h line 126, and if i shut it down, next one will be in line 342 shuting these both allows me to play it.

**Comment by kd-11** [comment]
Thanks. Can you check if enabling "disable FIFO reordering" makes the verification error go away?

**Comment by vietnam13231** [comment]
Infamous still unable to get to menu, black screen, FPS freezes, only sound goes.

**Comment by kd-11** [comment]
Infamous freezing should be fixed now

**Comment by digitaldude555** [comment]
screenshot 1026
yes

**Comment by Hasster1** [comment]
Yup, infamous is working.

**Comment by stride21** [comment]
Infamous 2 also had that issue and is now fixed.

**Comment by Anty0** [comment]
Current version of PR breaks R&C: ToD.
It was working fine in first versions of PR and works fine on master.

So far I know (tested twice, with same results):
- The ship lands under ground and lot of other stuff spawn under ground (= instant death). (Gets fixed after few respawns.)
- Can't switch planets: Freezes or crashes with message: RawSPU not found (cmd=0x40, lsa=0x2e9f9, ea=0xe68fb6f9, tag=0x3, size=0x40)

Log on PR (from second test) with crash: [RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/2577712/RPCS3.log.gz)
Log on master: [RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/2577717/RPCS3.log.gz)

PS: Log on master may be damaged. I was unable to safe exit RPCS3, when I finished testing. And I was to lazy to do it again...

**Comment by kd-11** [comment]
Pretty sure you didnt actually test latest master. The new builds are just rebased on the new commits recently added into master, one which is known to be broken. There are 2 PRs already submitted to fix it.

**Comment by Anty0** [comment]
Tested on (from log): RPCS3 v0.0.5-7496-0e0a82e5 Alpha | master
But only once, so maybe it was just "lucky run". :)
Anyway sorry for false alert...
I will test it with mentioned PRs.

**Comment by HexyFify** [comment]
fps in spyro has increased up to 75% in demanding areas

**Comment by pepesilvialin** [comment]
In yakuza kanzen, the FPS increased to 19-20 from 15-16 at the main street of the whore town with the lastest 58c517c build
image

**Comment by fifazalata** [comment]
In MGS Collection
sometimes it flashes then the emulator hangs with this:

`F {rsx::thread} RSX: class std::runtime_error thrown: Unreachable
(in file c:\projects\rpcs3\rpcs3\emu\rsx\rsxfifo.cpp:291)`

**Comment by kd-11** [comment]
@fifazalata Does the flashing still happen when FIFO reordering is disabled? Theoretically that path can only trigger if 2 END calls are called without a BEGIN in between them which is certainly strange.

**Comment by fifazalata** [comment]
when I set Disable FIFO Reordering: true, everything gets back to normal
the strange thing is: after I set it back again to false, everything is back to being normal again
So I don't know if this happens randomly in case it's set to the default (false)
but in any case, as of now, everything is working as it should when it's set to the default (false).

**Comment by kd-11** [comment]
There probably was a fifo desync before thing started getting out of hand, would have been nice to have a log of what happens just before the submission of the broken command stream. Pretty sure realhw would have also freaked out at that set of events.

**Comment by isshininu** [comment]
[RPCS3.zip](https://github.com/RPCS3/rpcs3/files/2595007/RPCS3.zip)
Demon's Souls: spams "RSX: Fault in uninterruptible code!" in log [only when using Vulkan rendderer]

**Comment by Xcedf** [comment]
Uncharted 1 has some random vertex exploading issues with game's surrounding environment on this PR, it started with some of the first three commits, can't tell more precise cause even if with compilation fixes they have no render just blackscreen

**Comment by Anty0** [comment]
R&C: ToD also spams that error on my device, but it didn't break anything, so I haven't reason to report that...

**Comment by isshininu** [comment]
> R&C: ToD also spams that error on my device, but it didn't break anything, so I haven't reason to report that...

But Demon's Souls spams shittons errors per second already, why more?

**Comment by Kravickas** [comment]
Well Crysis 1 trigger `F {rsx::thread} RSX: class std::runtime_error thrown: Unreachable (in file c:\projects\rpcs3\rpcs3\emu\rsx\rsxfifo.cpp:291)` without disabling FIFO reordering every time after few sec ingame, but that game is a mess so i didnt bother to tell.

**Comment by kd-11** [comment]
If you are able to reproduce consistently the verification error, attach full log file. The 'log spam' is both important and harmless, indicates caution due to faulting unexpectedly, a situation for which there is no easy solution. It can be downgraded to trace to avoid annoying end users but may make it harder to debug any regressions.

**Comment by altiereslima** [comment]
Minecraft: Performance is much worse.

**Comment by kd-11** [comment]
Well, according to other testers its much faster, so we'll need more information on what you're doing. Attach screenshots and logs of master vs this branch.

**Comment by kd-11** [comment]
Looks to be some kind of AMD issue, might look into it before merge. I'm still on 18.8,1 (vulkan version 2.0.45) but the small suite of samples tested does not show any significant regressions yet.

**Comment by kd-11** [comment]
nvm, i figured out the UC1 issue. Its a minor bug with the vertex cache.

**Comment by kd-11** [comment]
UC1 should be fixed with the latest commit; retest.

**Comment by Xcedf** [comment]
Also confirm, Uncharted fixed

**Comment by fifazalata** [comment]
@kd-11 the flashing with emulator freeze with this error is back again in MGS with latest commit
```
F {rsx::thread} RSX: class std::runtime_error thrown: Unreachable
(in file c:\projects\rpcs3\rpcs3\emu\rsx\rsxfifo.cpp:291)
```
Here's a log this time, hope it helps
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/2604495/RPCS3.log.gz)

**Comment by xYaroslavGTx** [comment]
screenshot_92

**Comment by SanderinBelgium** [comment]
not sure what i should test, please let me know cause i really like ps3 games and i wanna help. My pc: GTX1080ti Amd Ryzen 7 2700X (8cores, 16threads and base clocked at 3.7Ghz) and 32GB's of ram.
the games that i have: EA Skate (1,2 and 3) Uncharted 2 and 3, Gran Turismo 5, Red Dead Redemption and soon couple more cause i borrowed some to my friends and i also have like 20-60 more but can't get to them soon.

**Comment by SanderinBelgium** [comment]
I am playing in 4k with 16x anistropic filter

Skate 3:
 (it feels like there's less delay, before it felt like there was a delay between movements)
(no real framedrops but framerate doesn't go higher certain areas)
but still can't play the first mission not where you follow but do a trick

Skate 2
Still runs like trash (i didn't see any improvement

Red Dead Redemption:
16fps on main menu
10fps in cutscene
around 8 fps in-game

**Comment by hyperspeedgx** [comment]
I got one mem access violation with bc7d4b31, not sure if it is random yet.
image

**Comment by arabek** [comment]
Does not compile on linux currently:

`
$ ninja
[1/300] cd /home/arab/src/rpcs3_build/rpcs3 && /usr/bin/cmake -DSOURCE_DIR=/home/arab/src/rpcs3/rpcs3 -P /home/arab/src/rpcs3/rpcs3/git-version.cmake
-- Found Git: /usr/bin/git (found version "2.19.2") 
-- GIT_VERSION: 7537-23b5d39a
-- GIT_BRANCH: HEAD
[2/300] Building CXX object rpcs3/Emu/CMakeFiles/rpcs3_emu.dir/__/__/Utilities/mutex.cpp.o
FAILED: rpcs3/Emu/CMakeFiles/rpcs3_emu.dir/__/__/Utilities/mutex.cpp.o 
/usr/bin/c++  -DGLX_GLXEXT_PROTOTYPES -DGL_GLEXT_PROTOTYPES -DHAVE_ALSA -DHAVE_LIBEVDEV -DHAVE_PULSE -DHAVE_VULKAN -DLLVM_AVAILABLE -DNDEBUG -DVK_USE_PLATFORM_WAYLAND_KHR -I/home/arab/src/rpcs3/rpcs3 -I/home/arab/src/rpcs3 -I/home/arab/src/rpcs3/3rdparty/yaml-cpp/include -isystem /usr/include/libevdev-1.0 -I/home/arab/src/rpcs3/3rdparty/pugixml/src -I/home/arab/src/rpcs3/3rdparty/GSL/include -I/usr/include/AL -I/home/arab/src/rpcs3/3rdparty/stblib -I3rdparty/libpng -I/home/arab/src/rpcs3/3rdparty/libpng -I/home/arab/src/rpcs3/llvm/include -I3rdparty/llvm_build/include -I/home/arab/src/rpcs3/asmjitsrc/../asmjit/src -I/home/arab/src/rpcs3/3rdparty/ffmpeg/linux/x86_64/include -I/home/arab/src/rpcs3/3rdparty/cereal/include -I/home/arab/src/rpcs3/3rdparty/GL -I/home/arab/src/rpcs3/Vulkan/glslang/SPIRV/.. -I/home/arab/src/rpcs3/Vulkan/glslang/glslang/.. -I/home/arab/src/rpcs3/3rdparty/xxHash -Wno-invalid-pch -fexceptions -Wl,--exclude-libs,ALL -O3 -DNDEBUG   -Wno-attributes -Wno-enum-compare -Wno-invalid-offsetof -msse -msse2 -mcx16 -march=native -std=gnu++1z  -Winvalid-pch -include "/home/arab/src/rpcs3_build/rpcs3/Emu/cotire/rpcs3_emu_CXX_prefix.hxx" -MD -MT rpcs3/Emu/CMakeFiles/rpcs3_emu.dir/__/__/Utilities/mutex.cpp.o -MF rpcs3/Emu/CMakeFiles/rpcs3_emu.dir/__/__/Utilities/mutex.cpp.o.d -o rpcs3/Emu/CMakeFiles/rpcs3_emu.dir/__/__/Utilities/mutex.cpp.o -c /home/arab/src/rpcs3/Utilities/mutex.cpp
In file included from /home/arab/src/rpcs3/Utilities/mutex.cpp:2:0:
/home/arab/src/rpcs3/Utilities/sync.h: In instantiation of ‘balanced_wait_until(atomic_t<T>&, u64, Pred&&)::<lambda(T&, auto:1 ...)> [with auto:1 = {}; T = unsigned int; Pred = shared_mutex::imp_wait()::<lambda(u32&, auto:2, ...)>]’:
/home/arab/src/rpcs3/Utilities/sync.h:236:19:   required from ‘bool balanced_wait_until(atomic_t<T>&, u64, Pred&&) [with T = unsigned int; Pred = shared_mutex::imp_wait()::<lambda(u32&, auto:2, ...)>; u64 = long unsigned int]’
/home/arab/src/rpcs3/Utilities/mutex.cpp:56:3:   required from here
/home/arab/src/rpcs3/Utilities/sync.h:171:26: error: no matching function for call to ‘invoke(shared_mutex::imp_wait()::<lambda(u32&, auto:2, ...)>, unsigned int&)’
    auto ret = std::invoke(std::forward<Pred>(pred), _new, args...);
               ~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/arab/src/rpcs3/Utilities/Atomic.h:4:0,
                 from /home/arab/src/rpcs3/rpcs3/stdafx.h:29,
                 from /home/arab/src/rpcs3_build/rpcs3/Emu/cotire/rpcs3_emu_CXX_prefix.cxx:4,
                 from /home/arab/src/rpcs3_build/rpcs3/Emu/cotire/rpcs3_emu_CXX_prefix.hxx:4:
/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/g++-v7/functional:77:5: note: candidate: template<class _Callable, class ... _Args> std::invoke_result_t<_Callable, _Args ...> std::invoke(_Callable&&, _Args&& ...)
     invoke(_Callable&& __fn, _Args&&... __args)
     ^~~~~~
/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/g++-v7/functional:77:5: note:   template argument deduction/substitution failed:
In file included from /home/arab/src/rpcs3/Utilities/mutex.cpp:2:0:
/home/arab/src/rpcs3/Utilities/sync.h: In instantiation of ‘bool balanced_wait_until(atomic_t<T>&, u64, Pred&&) [with T = unsigned int; Pred = shared_mutex::imp_wait()::<lambda(u32&, auto:2, ...)>; u64 = long unsigned int]’:
/home/arab/src/rpcs3/Utilities/mutex.cpp:56:3:   required from here
/home/arab/src/rpcs3/Utilities/sync.h:236:19: error: could not convert ‘test_pred.balanced_wait_until(atomic_t<T>&, u64, Pred&&) [with T = unsigned int; Pred = shared_mutex::imp_wait()::<lambda(u32&, auto:2, ...)>; u64 = long unsigned int]::<lambda(unsigned int&, auto:1 ...)>(value)’ from ‘void’ to ‘bool’
  while (!test_pred(value))
          ~~~~~~~~~^~~~~~~
/home/arab/src/rpcs3/Utilities/sync.h:236:9: error: in argument to unary !
  while (!test_pred(value))
         ^~~~~~~~~~~~~~~~~
/home/arab/src/rpcs3/Utilities/sync.h:240:18: error: could not convert ‘test_pred.balanced_wait_until(atomic_t<T>&, u64, Pred&&) [with T = unsigned int; Pred = shared_mutex::imp_wait()::<lambda(u32&, auto:2, ...)>; u64 = long unsigned int]::<lambda(unsigned int&, auto:1 ...)>(value)’ from ‘void’ to ‘bool’
    if (!test_pred(value) && !test_pred(value, nullptr))
         ~~~~~~~~~^~~~~~~
/home/arab/src/rpcs3/Utilities/sync.h:240:8: error: in argument to unary !
    if (!test_pred(value) && !test_pred(value, nullptr))
        ^~~~~~~~~~~~~~~~~
In file included from /home/arab/src/rpcs3/Utilities/Atomic.h:4:0,
                 from /home/arab/src/rpcs3/rpcs3/stdafx.h:29,
                 from /home/arab/src/rpcs3_build/rpcs3/Emu/cotire/rpcs3_emu_CXX_prefix.cxx:4,
                 from /home/arab/src/rpcs3_build/rpcs3/Emu/cotire/rpcs3_emu_CXX_prefix.hxx:4:
/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/g++-v7/functional:77:5: error: ‘std::invoke_result_t<_Callable, _Args ...> std::invoke(_Callable&&, _Args&& ...) [with _Callable = shared_mutex::imp_wait()::<lambda(u32&, auto:2, ...)>; _Args = {unsigned int&, std::nullptr_t&}; std::invoke_result_t<_Callable, _Args ...> = bool]’, declared using local type ‘shared_mutex::imp_wait()::<lambda(u32&, auto:2, ...)>’, is used but never defined [-fpermissive]
     invoke(_Callable&& __fn, _Args&&... __args)
     ^~~~~~
/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/g++-v7/functional:77:5: warning: ‘std::invoke_result_t<_Callable, _Args ...> std::invoke(_Callable&&, _Args&& ...) noexcept (false) [with _Callable = shared_mutex::imp_wait()::<lambda(u32&, auto:2, ...)>; _Args = {unsigned int&, std::nullptr_t&}]’ used but never defined
[35/300] Building CXX object rpcs3/Emu/CMakeFiles/rpcs3_emu.dir/Cell/Modules/cellAdec.cpp.o
ninja: build stopped: subcommand failed.
`

**Comment by arabek** [comment]
> Does not compile on linux currently:

Disregard that. Master doesn't either.

**Comment by Anty0** [comment]
I think I have found regression.
In R&C Trilogy in R&C 1.
Can be reproduced like this:
- Start R&C 1
- Load saved game
- Game will freeze after ship land animation.

It seems like some kind of race conditions.

Happens every time using this PR with configuration I'm using on master. (Tested ~4 times.)
Can be fixed by enabling "Disable FIFO reordering" in debug.

I have never encountered this problem on master. (Tested on current master ~3 times.)

I'm on Arch Linux.
I have forgot to backup logs so I don't have them now, but in case it can help, I can get them.

**Comment by kd-11** [comment]
Report any regressions/issues [here](https://github.com/RPCS3/rpcs3/issues/5380). Remember to include screenshots and logs.

**Pull request merged**